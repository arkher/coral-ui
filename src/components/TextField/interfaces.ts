import { BoxProps } from '@shopify/restyle';
import { TextInput } from 'react-native';
import { Theme } from '../../themes/default';

type InputProps = {
  variant: 'small' | 'medium';
} & Partial<{
  label: string;
  assistiveText: string;
  placeholder: string;
  bw: CustomBorderWidth;
  p: CustomSpacing;
  status: 'success' | 'error' | undefined;
}> &
  Partial<BoxProps<Theme>>;

type InputRef = HTMLInputElement & TextInput;

export { InputProps, InputRef };
