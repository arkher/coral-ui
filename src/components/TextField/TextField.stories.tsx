import React from 'react';
import { withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { withDesign } from 'storybook-addon-designs';
import { StoriesView } from '../../stories/StorieView';
import Text from '../Text';
import TextField from './TextField';

export default {
  title: 'Componente/TextField',
  component: TextField,
  argTypes: {
    label: {
      control: { type: 'text' },
    },
    placeholder: {
      control: { type: 'text' },
    },
    assistiveText: {
      control: { type: 'text' },
    },
    status: {
      options: ['enable', 'success', 'error'],
      control: { type: 'inline-radio' },
    },
  },
  decorators: [withDesign],
};

interface TextInputProps {
  label: string;
  placeholder: string;
  assistiveText: string;
  // eslint-disable-next-line react/no-unused-prop-types
  status: 'success' | 'error' | undefined;
}

const urlHandoff =
  'https://www.figma.com/file/3raVfIADTUZCzFOOaQ9PMQ/HANDOFF-%7C-Core-Components-Mobile-%7C-Institucional?node-id=222%3A4012';

export const Small = ({
  label,
  placeholder,
  assistiveText,
  status,
}: TextInputProps): React.ReactNode => (
  <TextField
    label={label}
    assistiveText={assistiveText}
    placeholder={placeholder}
    variant="small"
    status={status}
  />
);

Small.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const Medium = ({
  label,
  placeholder,
  assistiveText,
}: TextInputProps): React.ReactNode => (
  <TextField
    label={label}
    assistiveText={assistiveText}
    placeholder={placeholder}
    variant="medium"
  />
);

Medium.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

// Add all stories to RN/Expo storybook
storiesOf('TextField-ds', module)
  .addDecorator(getStory => <StoriesView>{getStory()}</StoriesView>)
  .addDecorator(withKnobs)
  .add('Small', () => (
    <>
      <Text mb="sm">TextField / Success / Error</Text>

      <TextField
        label="Label"
        assistiveText="Texto de suporte"
        placeholder="Placeholder"
        variant="small"
      />

      <TextField
        label="Label"
        assistiveText="Texto de suporte"
        placeholder="Placeholder"
        variant="small"
        status="success"
      />

      <TextField
        label="Label"
        assistiveText="Texto de suporte"
        placeholder="Placeholder"
        variant="small"
        status="error"
      />
    </>
  ))
  .add('Medium', () => (
    <>
      <Text mb="sm">TextField / Success / Error</Text>

      <TextField
        label="label"
        assistiveText="Texto de suporte"
        placeholder="Placeholder"
        variant="medium"
      />

      <TextField
        label="Label"
        assistiveText="Texto de suporte"
        placeholder="Placeholder"
        variant="medium"
        status="success"
      />

      <TextField
        label="Label"
        assistiveText="Texto de suporte"
        placeholder="Placeholder"
        variant="medium"
        status="error"
      />
    </>
  ));
