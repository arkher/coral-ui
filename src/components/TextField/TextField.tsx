import React, { useEffect, useRef, useState } from 'react';
import { SafeAreaView } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { useTheme } from '@shopify/restyle';
import { Theme } from '../../themes/institucional';

import Text from '../Text';
import Box from '../Box';
import Input from '../Input';
import { TextFieldProps, TypeVariantHeight } from './interfaces';
import { InputFowardEvents } from '../Input/interfaces';

const TextField: React.FC<TextFieldProps> = ({
  label,
  placeholder,
  variant,
  assistiveText,
  status,
  keyboardType,
  autoCapitalize,
  returnKeyType,
  value,
  onChange,
}) => {
  const { colors } = useTheme<Theme>();
  const textfieldRef = useRef<InputFowardEvents>(null);

  const statusKeyPair = {
    error: colors['feedback-error-base'],
    success: colors['feedback-success-base'],
    default: colors['neutral-dark'],
  };

  const variantHeight: TypeVariantHeight = {
    small: 'xs',
    medium: 'sm',
  };

  useEffect(() => {
    if (status === 'error') {
      textfieldRef.current?.error();
    }

    if (status === 'success') {
      textfieldRef.current?.success();
    }
  }, [status]);

  return (
    <SafeAreaView>
      {!!label && (
        <Text fs="md" fontWeight="700" color="neutral-darkest">
          {label}
        </Text>
      )}
      <Input
        ref={textfieldRef}
        placeholder={placeholder}
        variant={variantHeight[variant]}
        keyboardType={keyboardType}
        autoCapitalize={autoCapitalize}
        returnKeyType={returnKeyType}
        value={value}
        onChange={e => {
          onChange && onChange({ ...e, current: textfieldRef.current });
        }}
        my="quark"
        px="xs"
        icon="close"
        style={{
          flex: 1,
        }}
      />
      {!!assistiveText && (
        <Box flexDirection="row" alignItems="center">
          {status === 'success' && (
            <Icon
              name="check-circle-outline"
              size={24}
              color={statusKeyPair[status]}
            />
          )}
          {status === 'error' && (
            <Icon
              name="alert-circle-outline"
              size={24}
              color={statusKeyPair[status]}
            />
          )}
          <Text ml="quark" fs="sm" color="neutral-darkest">
            {assistiveText}
          </Text>
        </Box>
      )}
    </SafeAreaView>
  );
};

export default TextField;
