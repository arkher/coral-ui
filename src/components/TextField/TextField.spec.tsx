import React from 'react';
import { cleanup, render } from '@testing-library/react-native';
import { ThemeProvider } from '@shopify/restyle';
import TextField from './TextField';
import themeMaestro from '../../themes/maestro';

afterEach(cleanup);

jest.mock('react-native-vector-icons/MaterialCommunityIcons', () => 'Icon');

test('should have a placeholder passed by prop', () => {
  const { getAllByPlaceholderText } = render(
    <ThemeProvider theme={themeMaestro}>
      <TextField
        variant="medium"
        placeholder="Placeholder"
        assistiveText="Texto de suporte"
      />
    </ThemeProvider>,
  );

  expect(getAllByPlaceholderText('Placeholder').length).toBe(1);
});

test('should have a label passed by prop', () => {
  const { getAllByText } = render(
    <ThemeProvider theme={themeMaestro}>
      <TextField
        label="Label do TextArea"
        variant="medium"
        placeholder="Placeholder"
      />
    </ThemeProvider>,
  );

  expect(getAllByText('Label do TextArea').length).toBe(1);
});

test('should have a success feedback', () => {
  const { toJSON } = render(
    <ThemeProvider theme={themeMaestro}>
      <TextField
        label="Label do TextArea"
        variant="medium"
        status="success"
        placeholder="Placeholder"
      />
    </ThemeProvider>,
  );

  expect(toJSON()).toMatchSnapshot();
});

test('should have a error feedback', () => {
  const { toJSON } = render(
    <ThemeProvider theme={themeMaestro}>
      <TextField
        label="Label do TextArea"
        variant="medium"
        status="error"
        placeholder="Placeholder"
      />
    </ThemeProvider>,
  );

  expect(toJSON()).toMatchSnapshot();
});

test('should TextField render correctly', async () => {
  const { toJSON } = render(
    <ThemeProvider theme={themeMaestro}>
      <TextField variant="medium" placeholder="Placeholder" />
    </ThemeProvider>,
  );

  expect(toJSON()).toMatchSnapshot();
});
