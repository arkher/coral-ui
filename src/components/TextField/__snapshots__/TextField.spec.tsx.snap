// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should TextField render correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    onChange={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#5D6A75",
          "borderRadius": 4,
          "display": "flex",
          "flexDirection": "row",
          "marginVertical": 4,
          "paddingHorizontal": 16,
        },
        Array [
          Object {
            "borderWidth": 1,
            "height": 56,
          },
        ],
      ]
    }
    testID="Box"
  >
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Placeholder"
      placeholderTextColor="#5D6A75"
      rejectResponderTermination={true}
      selectionColor="#5D6A75"
      style={
        Array [
          Object {},
          Object {
            "flex": 1,
          },
        ]
      }
      testID="Input"
      underlineColorAndroid="transparent"
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "display": "flex",
            "marginLeft": 4,
          },
          Array [
            Object {},
          ],
        ]
      }
      testID="Box"
    >
      <Icon
        color="neutral-dark"
        name="close"
        size={24}
      />
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`should have a error feedback 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <Text
    style={
      Array [
        Object {
          "color": "#2D3338",
          "display": "flex",
          "fontFamily": "\\"Inter\\", sans-serif",
          "fontWeight": "700",
        },
        Array [
          Object {
            "fontSize": 14,
          },
        ],
      ]
    }
  >
    Label do TextArea
  </Text>
  <View
    onChange={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#E20A0A",
          "borderRadius": 4,
          "display": "flex",
          "flexDirection": "row",
          "marginVertical": 4,
          "padding": 8,
        },
        Array [
          Object {
            "borderWidth": 2,
            "height": 56,
          },
        ],
      ]
    }
    testID="Box"
  >
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Placeholder"
      placeholderTextColor="#5D6A75"
      rejectResponderTermination={true}
      selectionColor="#5D6A75"
      style={
        Array [
          Object {},
          Object {
            "flex": 1,
          },
        ]
      }
      testID="Input"
      underlineColorAndroid="transparent"
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "display": "flex",
            "marginLeft": 4,
          },
          Array [
            Object {},
          ],
        ]
      }
      testID="Box"
    >
      <Icon
        color="neutral-dark"
        name="close"
        size={24}
      />
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`should have a success feedback 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <Text
    style={
      Array [
        Object {
          "color": "#2D3338",
          "display": "flex",
          "fontFamily": "\\"Inter\\", sans-serif",
          "fontWeight": "700",
        },
        Array [
          Object {
            "fontSize": 14,
          },
        ],
      ]
    }
  >
    Label do TextArea
  </Text>
  <View
    onChange={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#00B649",
          "borderRadius": 4,
          "display": "flex",
          "flexDirection": "row",
          "marginVertical": 4,
          "padding": 8,
        },
        Array [
          Object {
            "borderWidth": 2,
            "height": 56,
          },
        ],
      ]
    }
    testID="Box"
  >
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Placeholder"
      placeholderTextColor="#5D6A75"
      rejectResponderTermination={true}
      selectionColor="#5D6A75"
      style={
        Array [
          Object {},
          Object {
            "flex": 1,
          },
        ]
      }
      testID="Input"
      underlineColorAndroid="transparent"
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "display": "flex",
            "marginLeft": 4,
          },
          Array [
            Object {},
          ],
        ]
      }
      testID="Box"
    >
      <Icon
        color="neutral-dark"
        name="close"
        size={24}
      />
    </View>
  </View>
</RCTSafeAreaView>
`;
